src/arraymancer/linear_algebra/special_matrices.nim

```diff
@@ -47,3 +47,137 @@ proc vandermonde*[T](x: Tensor[T], order: int): Tensor[float] =
+  ## Result:
+  ##      - The constructed, square diagonal matrix
+  assert d.rank == 1
+  let size = d.size + abs(k)
+  result = zeros[T](size, size)
+
+  if k >= 0:
+    for i in 0 ..< d.size:
+      result[i, i+k] = d[i]
+  else:
+    for i in 0 ..< d.size:
+      result[i-k, i] = d[i]
+
+proc diagonal*[T](a: Tensor[T], k = 0, anti = false): Tensor[T] {.noInit.}=
+  ## Gets the k-th diagonal (or anti-diagonal) of a matrix
+  ##
+  ## Input:
+  ##      - A matrix (which can be rectangular)
+  ##      - k: The index k of the diagonal that will be extracted. The default is 0 (i.e. the main diagonal).
+  ##        Use k>0 for diagonals above the main diagonal, and k<0 for diagonals below the main diagonal.
+  ##      - anti: If true, get the k-th "anti-diagonal" instead of the k-th regular diagonal.
+  ## Result:
+  ##      - A copy of the diagonal elements as a rank-1 tensor
+  assert a.rank == 2
+  assert k < a.shape[0], &"Diagonal index ({k=}) exceeds matrix height ({a.shape[0]})"
+  assert k < a.shape[1], &"Diagonal index ({k=}) exceeds matrix width ({a.shape[1]})"
+  let size = min(a.shape[0], a.shape[1]) - abs(k)
+  result = newTensor[T](size)
+
+  if anti:
+    if k >= 0:
+      let size = min(a.shape[0] - abs(k), a.shape[1])
+      result = newTensor[T](size)
+      for i in 0 ..< size:
+        result[i] = a[a.shape[0]-1-(i+k), i]
+    else:
+      let size = min(a.shape[0], a.shape[1] - abs(k))
+      result = newTensor[T](size)
+      for i in 0 ..< size:
+        result[i] = a[a.shape[0]-1-i, i-k]
+  else:
+    if k >= 0:
+      let size = min(a.shape[0], a.shape[1] - abs(k))
+      result = newTensor[T](size)
+      for i in 0 ..< size:
+        result[i] = a[i, i+k]
+    else:
+      let size = min(a.shape[0]-abs(k), a.shape[1])
+      result = newTensor[T](size)
+      for i in 0 ..< size:
+        result[i] = a[i-k, i]
+
+proc set_diagonal*[T](a: var Tensor[T], d: Tensor[T], k = 0, anti = false) =
+  ## Sets a diagonal of a matrix (in place)
+  ##
+  ## Input:
+  ##      - The matrix that will be changed in place.
+  ##      - Rank-1 tensor containg the elements that will be copied into the selected diagonal.
+  ##      - k: The index k of the diagonal that will be changed. The default is 0 (i.e. the main diagonal).
+  ##        Use k>0 for diagonals above the main diagonal, and k<0 for diagonals below the main diagonal.
+  ##      - anti: If true, set the k-th "anti-diagonal" instead of the k-th regular diagonal.
+  assert a.rank == 2
+  assert d.rank == 1
+  assert k < a.shape[0], &"Diagonal index ({k=}) exceeds matrix height ({a.shape[0]})"
+  assert k < a.shape[1], &"Diagonal index ({k=}) exceeds matrix width ({a.shape[1]})"
+  if anti:
+    if k >= 0:
+      when compileOption("boundChecks"):
+        let size = min(a.shape[0] - abs(k), a.shape[1])
+        assert size == d.size, &"Input size ({d.size}) does not match the {k}-th upper anti-diagonal size ({size})"
```

Comment:
I'm not sure if this is what we would "intend". For example if the user wishes bound checks on a `danger` build the assertion is still going to be ignored. If we explicitly activate the assertion only for bounds checks being active, then maybe we should use `doAssert` to also have it work on `danger` builds or use a regular exception? Same for the other `when` below of course.

Reply:
I will replace the calls to assert with doAssert as you suggest.